### PR TITLE
MODULES-3360 Fix correct catalina_home handling

### PR DESCRIPTION
**Pull request by dploeger**
This fixes MODULES-3360

### CODE REVIEW

**Comment by jonnytpuppet** [comment]
Hi There

Thanks for your contribution!

In order for us to merge this, it would be appreciated if you could have a look at why the tests are failing.

Thanks

**Comment by dploeger** [comment]
Hmm... I don't think, that my changes resulted in that error, because they were very minor. Could you please have a look? Thanks.

File: manifests/install.pp [hunk 1]
```diff
@@ -17,17 +17,18 @@
   $package_options        = undef,
 ) {
   include tomcat
+  $_catalina_home = pick($catalina_home, $::tomcat::catalina_home, $name)
```
Comment (bmjen) [inline manifests/install.pp]: Hi @dploeger.. this would be backwards incompatible, because the pick will always chose the value from `$::tomcat::catalina_home` which is defaulted to the value set in params.pp.

Comment (dploeger) [inline manifests/install.pp]: Yes, @bmjen, I guess you're right. However, I would've expect that behavior: If I set the default using the class I would want tomcat::install to use that, and not the name from tomcat::install itself, but opinions might differ on this one.

Finally, it's up to you. I don't know a backwards compatible way here, because switching $name and $::tomcat::catalina_home in that statement would also not work, because $name is always set.

Comment (bmjen) [inline manifests/install.pp]: I agree in that the catalina_home set by the main class should be the default. However, in doing so.. we will have to do a major version bump. I can keep this PR around with an appropriate milestone label.

**Comment by tphoney** [comment]
Apologies, i had to close then reopen this PR, to force TravisCI to pick up the reduced test run for docker. This meant that the test runs would sometimes fail through no fault of the author. The TravisCI tests now complete in 15 minutes.

File: manifests/install.pp (hunk 1 above)
Comment (hunner) [inline manifests/install.pp]: `$::tomcat::catalina_home` always has a value in the current master branch which will make `$name` never choosable.

**Comment by hunner** [comment]
tl;dr the base class is an ugly wart that shouldn't be used, but can't be cleanly removed yet.

When revising the usage of the tomcat module, I was picturing one `tomcat::install` declared per `catalina_home` and then one `tomcat::instance` declared per `catalina_base` and pointing at the desired `catalina_home` for the installation files. This would allow multiple installs each with multiple instances and low possibility of conflicting.

The fact that the base class exists and is referenced by the defined types is more of a fluke due to the module conflating catalina home and base previously and I had always planned to rework this code during a major bump. There is no way to reconcile the differences between "multi-instance" and "defaults for everything" as far as I can see though.

1. I could just remove the base class, so an install/instance must be declared per install/instance, but perhaps that would be frustrating for anyone who wants to just get started.
2. I could make the base class declare a "default" install/instance of tomcat, but really this is just to make getting started easier. This is not a workable solution for any production-worthy system.

For 2.0.0, I'm not going to break any of this, because I think the docs and usage pattern haven't shifted away from using the base class and I haven't done any deprecations yet.

I did think about deprecating declaring a `tomcat::instance` that did not have a corresponding `tomcat::install` (see this whole bit of uglyness: https://github.com/puppetlabs/puppetlabs-tomcat/blob/1.7.0/manifests/instance.pp#L59-L67). But there is something nice about being able to declare a `tomcat::instance` and be done with it, so I left it to be a future decision when I wrote https://github.com/puppetlabs/puppetlabs-tomcat/pull/125 and so it remains.

**Comment by hunner** [comment]
I forgot to mention that my vision for the base class, install, and instance looks something like how the base class in my wordpress module declares an instance with simple defaults, but the real power comes from understanding how to declare instances yourself: https://github.com/hunner/puppet-wordpress/blob/master/manifests/init.pp vs https://github.com/hunner/puppet-wordpress/blob/master/manifests/instance.pp
